### PR TITLE
Refactor list filters

### PR DESCRIPTION
**Pull request by alterphp**
Replaces #83.

### CODE REVIEW

**Comment by alterphp** [comment]
Hi @CheapHasz !

I've pushed some modifications last friday. I'd like to get your feedback about them.

File: src/Model/ListFilter.php [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+    public const OPERATOR_IN = 'in';
+    public const OPERATOR_NOTIN = 'notin';
+    public const OPERATOR_GT = 'gt';
+    public const OPERATOR_GTE = 'gte';
+    public const OPERATOR_LT = 'lt';
+    public const OPERATOR_LTE = 'lte';
+
+    private static $operatorValues = null;
+
+    public static function createFromRequest(string $property, $operator, $value)
```
[the remark this answers is not in the document]
Comment (alterphp) [reply src/Model/ListFilter.php]: sure !

File: src/EventListener/PostQueryBuilderSubscriber.php [hunk 2]
```diff
@@ -114,74 +109,118 @@ protected function applyRequestFilters(QueryBuilder $queryBuilder, array $filter
...
-            case \is_array($value):
+        $value = $this->filterEasyadminAutocompleteValue($listFilter->getValue());
+        // Empty string and numeric keys is considered as "not applied filter"
+        if (null === $value || '' === $value || \is_int($field)) {
+            return;
+        }
+
+        // Add root entity alias if none provided
+        $queryField = $listFilter->getProperty();
+        if (false === \strpos($queryField, '.')) {
```
[the remark this answers is not in the document]
Comment (alterphp) [reply src/EventListener/PostQueryBuilderSubscriber.php]: This is intended not to prefix related entities properties by the root alias.

Imagine you have overriden an admin Product Controller to leftJoin with its related Category :

```
    protected function createListQueryBuilder($entityClass, $sortDirection, $sortField = null, $dqlFilter = null)
    {
        $qb = parent::createListQueryBuilder($entityClass, $sortDirection, $sortField, $dqlFilter);

        $qb
            ->addSelect('cat')
            ->leftJoin('entity.category', 'cat')
        ;

        return $qb;
    }
```

This allows to display a field `cat.name` in the list without having X more queries (1 for each displaid row).

And you can define a filter on property `cat.season` for example, to filter products on their category season. DQL only expects one dot separation in properties (`alias`.`field`) thats why we have to check this before prefixing with root alias.

File: src/Model/ListFilter.php [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+     * Returns operators list.
+     *
+     * @return array
+     */
+    public static function getOperatorsList()
+    {
+        // Build $operatorValues if this is the first call
+        if (null === static::$operatorValues) {
+            static::$operatorValues = [];
+            $refClass = new \ReflectionClass(\get_called_class());
```
[the remark this answers is not in the document]
Comment (alterphp) [reply src/Model/ListFilter.php]: No good reason, that code block comes from an old library of mine that has been written with PHP 5.4. I will change for `static::class`.

File: src/EventListener/PostQueryBuilderSubscriber.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (CheapHasz) [reply src/EventListener/PostQueryBuilderSubscriber.php]: alright

**Pull request merged**